File: 152.py
class Solution(object):
    def maxProduct(self, nums):
        ans = nums[0]  # Initialize with the first element of nums
        cmax, cmin = 1, 1  # Initializing cmax and cmin

        # Start from the first element (nums[0])
        for i in nums:
            temp = i * cmax  # Store the product of current max and num
            cmax = max(i * cmax, i * cmin, i)  # Update current max
            cmin = min(temp, i * cmin, i)  # Update current min
            ans = max(ans, cmax)  # Update the result with the max value

        return ans  # Return the final answer

File: 152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans=nums[0];
        int cmax=1, cmin=1;
        for(int i=0;i<nums.size();i++){
            int temp= cmax*nums[i];
            cmax=max(max(cmax*nums[i],cmin*nums[i]),nums[i]);
            cmin=min(min(temp,cmin*nums[i]),nums[i]);
            ans=max(ans, cmax);
        }
        return ans;
    }
};
